cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -474,17 +460,59 @@ impl YaccParser {
             i = k;
         }
 
+        // Next, the '(' pattern : type, ... ')'
         i = self.parse_ws(i, false)?;
-        let (i, pat) = self.parse_to_single_colon(i)?;
-        let (i, typ) = self.parse_to_eol(i + ':'.len_utf8())?;
-        let pat = pat.trim_start().trim_end().to_string();
-        let typ = typ.trim_start().trim_end().to_string();
-        if !(pat.is_empty() || typ.is_empty()) {
-            self.ast.parse_param_bindings = Some((pat, typ));
+        if let Some(mut j) = self.lookahead_is("(", i) {
+            let mut bindings: Vec<(String, String)> = Vec::new();
+
+            while j < self.src.len() {
+                // Some binding name, or pattern.
+                let k = self.parse_ws(j, false)?;
+                let (k, binding) = self.parse_to_single_colon(k)?;
+                let (k, typ, done) = self.parse_param_rust_type(k + ':'.len_utf8())?;
+                j = k;
+                bindings.push((binding.trim_end().to_string(), typ));
+                if done == true {
+                    break;
+                }
+            }
+            if !bindings.is_empty() {
+                self.ast.parse_param_bindings = Some(bindings);
+            }
+            i = j;
         }
         Ok(self.parse_ws(i, true)?)
     }
 
+    fn parse_param_rust_type(&mut self, i: usize) -> YaccResult<(usize, String, bool)> {
+        let mut j = i;
+        let mut brace_count = 0;
+
+        while j < self.src.len() {
+            let c = self.src[j..].chars().next().unwrap();
+            match c {
+                '\n' | '\r' => return Err(self.mk_error(YaccParserErrorKind::ReachedEOL, j)),
+                ')' | ',' if brace_count == 0 => {
+                    return Ok((
+                        j + c.len_utf8(),
+                        self.src[i..j].trim_start().trim_end().to_string(),
+                        c == ')',
+                    ));
+                }
+                '(' | '{' | '[' | '<' => {
+                    brace_count += 1;
+                    j += c.len_utf8();
+                }
+                ')' | '}' | '>' | ']' if brace_count > 0 => {
+                    brace_count -= 1;
+                    j += c.len_utf8();
+                }
```

Comment:
Do we need to handle whitespace explicitly here? Or perhaps it's just OK to pass it through to `rustc`?

Reply:
IIUC that this is about the whitespace trimming,
It is fine to remove it and pass it through to rust, I was mostly trimming it so the tests don't run into it.

Reply:
I wondered if we needed to manually trim it; but if what's here works as-is, that's fine with me!

Reply:
I managed to get rid of the `trim_start()`, but the trim_end() isn't so easy.